src/archives.c

```diff
@@ -0,0 +1,173 @@
+/*
+ *   Software Updater - client side
+ *
+ *      Copyright © 2012-2017 Intel Corporation.
+ *
+ *   This program is free software: you can redistribute it and/or modify
+ *   it under the terms of the GNU General Public License as published by
+ *   the Free Software Foundation, version 2 or later of the License.
+ *
+ *   This program is distributed in the hope that it will be useful,
+ *   but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *   GNU General Public License for more details.
+ *
+ *   You should have received a copy of the GNU General Public License
+ *   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ *   Authors:
+ *         Matthew Johnson <mathew.johnson@intel.com>
+ *
+ */
+
+#include <archive.h>
+#include <archive_entry.h>
+#include <stdlib.h>
+
+#include "swupd.h"
+
+/* _archive_check_err(ar, ret)
+ *
+ * Check and print archive errors */
+static int _archive_check_err(struct archive *ar, int ret)
+{
+	int is_fatal = 0;
+
+	if (ret < ARCHIVE_WARN || ret == ARCHIVE_RETRY) {
+		/* error was worse than a warning or error was ARCHIVE_RETRY (greater
+		 * than ARCHIVE_WARN), which indicates failure with retry possible.
+		 * Regardless, indicate that the operation did not succeed.  */
+		fprintf(stderr, "Error: %s\n", archive_error_string(ar));
+		is_fatal = 1;
+	} else if (ret < ARCHIVE_OK) {
+		/* operation succeeded, warning encountered */
+		fprintf(stderr, "Warning: %s\n", archive_error_string(ar));
+		is_fatal = 0;
+	}
+
+	return is_fatal;
+}
+
+/* copy_data(ar, aw)
+ *
+ * Copy archive data from ar to aw */
+static int copy_data(struct archive *ar, struct archive *aw)
+{
+	int r;
+	const void *buffer;
+	size_t size;
+	off_t offset;
+
+	for (;;) {
+		r = archive_read_data_block(ar, &buffer, &size, &offset);
+		if (r == ARCHIVE_EOF) {
+			return ARCHIVE_OK;
+		} else if (r < ARCHIVE_OK) {
+			return r;
+		}
+
+		r = archive_write_data_block(aw, buffer, size, offset);
+		if (r < ARCHIVE_OK) {
+			return r;
+		}
+	}
+}
+
+/* extract_to(tarfile, outputdir)
+ *
+ * Extracts tar archive tarfile to outputdir. Refuses to extract any object
+ * whose final location would be altered by a symlink on disk. */
+int extract_to(const char *tarfile, const char *outputdir)
+{
+	struct archive *a, *ext;
+	struct archive_entry *entry;
+	int flags;
+	int r = 0;
+
+	/* set which attributes we want to restore */
+	flags = ARCHIVE_EXTRACT_TIME;
+	flags |= ARCHIVE_EXTRACT_PERM;
+	flags |= ARCHIVE_EXTRACT_OWNER;
+	flags |= ARCHIVE_EXTRACT_XATTR;
+
+	/* set security flags */
+	flags |= ARCHIVE_EXTRACT_SECURE_SYMLINKS;
+	flags |= ARCHIVE_EXTRACT_SECURE_NODOTDOT;
+
+	/* set up read */
+	a = archive_read_new();
+	r = archive_read_support_format_all(a);
```

Comment:
You can save some CPU cycles by limiting the format to what will actually be present in the archive, i.e. tar.

Reply:
ah, you're right. I'll do that.

Reply:
Ah, I thought it was deliberate that you were being flexible.

Reply:
the flexible part is `archive_read_support_filter_all`